### PR TITLE
mt-gateway: support ingesting data into kafka

### PR DESCRIPTION
**Pull request by fitzoh**
Initial portion of this PR was porting the `github.com/raintank/tsdb-gw/publish` package to this repository for use in a new HTTP proxy (see https://github.com/grafana/metrictank-ops/issues/542 if you have access).

The portion after [this comment](https://github.com/grafana/metrictank/pull/1608#issuecomment-575830102) (specifically after the force push) pulls in the actual API endpoint and wires it up to Kafka.

A couple design decisions:
* This is intended to be a simplified API, so rate limiting functionality has been removed.
* As there is no concept of authentication/authorization within metrictank, the `isAdmin=true` path is always taken.

Once again, the latter portion of this PR is intended to be reviewed commit by commit.

Open questions:
* Do we need to consider any licensing/attribution issues?
    * probably fine, could use a +1 from @woodsaj ✔️
* Is there any dead code we can get rid of post-import?
   * removed ✔️
* Should we maintain the instrumentation present in the original code?
    * leave it ✔️
* Testing? Do we add this to a docker config? Which one?
    * toss it in `docker-dev-custom-cfg-kafka` alongside `tsdb-gw` for now

### CODE REVIEW

File: publish/publish.go
Comment (fitzoh) [inline publish/publish.go]: It doesn't look like this `Persistor` is actually used (either in this PR or anywhere in the tsdb-gw project), which I believe means we can drop it as well as the `metrics_client` package which was pulled in specifically for this.

File: publish/kafka/schemas.go
Comment (fitzoh) [inline publish/kafka/schemas.go]: This function and file can probably go away

File: publish/publish.go
Comment (replay) [reply publish/publish.go]: hmm i have a feeling that this was added because historically we used to use tsdb-gw also to ingest data into cortex, but now we don't do that anymore afaik

Comment (replay) [reply publish/publish.go]: so `metrics_client` is only necessary for the `persistor`? in that case i think we can remove all of it probably

File: publish/kafka/schemas.go
Comment (replay) [reply publish/kafka/schemas.go]: why can this go away? don't we need this when initializing the `kafka.mtPublisher`, to set its `schemas` property (https://github.com/grafana/metrictank/pull/1608/files#diff-c1de7345bb1a47b5638ddad84e8ae8eeR200)?

Comment (fitzoh) [reply publish/kafka/schemas.go]: Screen Shot 2020-01-16 at 1 12 21 PM

File: publish/publish.go
Comment (fitzoh) [reply publish/publish.go]: 3af36e930d997bc3facbc9d115d4ba5401c00ed5

File: publish/kafka/schemas.go
Comment (fitzoh) [reply publish/kafka/schemas.go]: removed in 34f4d3884938feb1e9b80211561a684afa5f1379

**Comment by replay** [comment]
To be sure I understand this diff right, this doesn't change the Sarama version, it only adds another package right? I'm asking because when changing Sarama versions it's usually a good idea to do some benchmarks to verify that there are no regressions.

**Comment by replay** [comment]
And currently none of this code which gets added is actually being used, right? That's all going to be used by your new http gateway?

**Comment by replay** [comment]
By merging the dependencies of the mt-gateway into the MT repo separately before they actually get used, isn't there a risk that we accidentally copy more stuff from tsdb-gw than what's actually going to be used and this will then be left there as dead code, or are you sure that all of this is needed?

**Comment by robert-milan** [comment]
> To be sure I understand this diff right, this doesn't change the Sarama version, it only adds another package right? I'm asking because when changing Sarama versions it's usually a good idea to do some benchmarks to verify that there are no regressions.

Yes that is correct. I thought the same thing at first glance, but it only adds `mocks`.

I understand that breaking it up like this makes for smaller and easier to read PRs, but in this case it does seem a little bit awkward. I think this should be part of the PR that uses this code in the HTTP proxy. Since #1602 is already merged, can we just rename this PR and add the HTTP proxy changes here as well? It is easy enough to review commit by commit.

> Do we need to consider any licensing/attribution issues?

We probably should take a look at everything

> Are we able to omit any portion of this package?

That would be easier to determine once we see some of the implementation

> Should we maintain the instrumentation present in the original?

I don't think we need to maintain the instrumentation unless there is a good reason for it.

**Comment by fitzoh** [comment]
>>To be sure I understand this diff right, this doesn't change the Sarama version, it only adds another package right? I'm asking because when changing Sarama versions it's usually a good idea to do some benchmarks to verify that there are no regressions.

>Yes that is correct. I thought the same thing at first glance, but it only adds mocks.

Correct, @robert-milan  


>And currently none of this code which gets added is actually being used, right? That's all going to be used by your new http gateway?

>I understand that breaking it up like this makes for smaller and easier to read PRs, but in this case it does seem a little bit awkward. I think this should be part of the PR that uses this code in the HTTP proxy. Since #1602 is already merged, can we just rename this PR and add the HTTP proxy changes here as well? It is easy enough to review commit by commit.


Yeah, good call.  This was mostly an attempt to unblock myself while #1602 was waiting on reviews... now that it's merged, I'll rebase and work on wiring things up wire things up

File: cmd/mt-gateway/ingest/metrics.go
Comment (replay) [inline cmd/mt-gateway/ingest/metrics.go]: there's a repeating pattern of writing a status code header, then writing a message on the following line, and in some cases also log a line with the same msg. 
would it make sense to put these into a small helper like `writeStatus(w, 499, "request cancelled", true)` (the last param could determine whether to log).

File: cmd/mt-gateway/api.go
Comment (replay) [inline cmd/mt-gateway/api.go]: that's just a convention: we usually sort the imports so that there are first all the standard library imports, then an empty line, then all the imports of other pkgs. (f.e. https://github.com/grafana/metrictank/blob/master/idx/memory/memory.go#L3)

File: cmd/mt-gateway/api.go
Comment (replay) [inline cmd/mt-gateway/api.go]: it might be helpful to new users if the message mentions that metrics ingestion is not enabled because no kafka brokers were configured.

File: cmd/mt-gateway/api.go
Comment (replay) [inline cmd/mt-gateway/api.go]: this message seems wrong

**Review by replay — CHANGES_REQUESTED** [review]
Looks good, just a few minor comments

File: cmd/mt-gateway/api.go
Comment (fitzoh) [reply cmd/mt-gateway/api.go]: 87cb16c62265e5574e616489cd4489007c8f53e6

File: cmd/mt-gateway/api.go
Comment (fitzoh) [reply cmd/mt-gateway/api.go]: 87cb16c62265e5574e616489cd4489007c8f53e6

File: cmd/mt-gateway/api.go
Comment (fitzoh) [reply cmd/mt-gateway/api.go]: 186fab2a2c2d42fa4de174c437a2f11361fb69c2

File: cmd/mt-gateway/ingest/metrics.go
Comment (fitzoh) [reply cmd/mt-gateway/ingest/metrics.go]: Yeah, that had crossed my mind.
Skipped the  boolean and just always log @ error (successful ingestions take a different code path)

1abc592486eb7c50c4d2b25b84f73d6c037704dc

File: cmd/mt-gateway/ingest/metrics.go
Comment (replay) [inline cmd/mt-gateway/ingest/metrics.go]: this could be `fmt.Fprint()` now, instead of `Fprintf`

Comment (fitzoh) [reply cmd/mt-gateway/ingest/metrics.go]: woops, 3237276fcfbf60999ddfa8328c93db1abb686e9c

**Review by replay — APPROVED** [review]
LGTM

**Pull request merged**
